src/Processors/Transforms/Streaming/JoinTransformWithAlignment.h

```diff
@@ -27,81 +30,120 @@ class JoinTransformWithAlignment final : public IProcessor
     Status prepare() override;
     void work() override;
 
+    void checkpoint(CheckpointContextPtr ckpt_ctx) override;
+    void recover(CheckpointContextPtr ckpt_ctx) override;
+
     static Block transformHeader(Block header, const HashJoinPtr & join);
 
 private:
-    Status prepareLeftInput();
-    Status prepareRightInput();
-
-    inline Int64 getRightWatermark(const Chunk & chunk) const
+    struct InputPortWithData
     {
-        return getWatermark(chunk, right_watermark_column_position, right_watermark_column_type);
-    }
+        explicit InputPortWithData(InputPort * input_port_) : input_port(input_port_) { }
 
-    inline Int64 getLeftWatermark(const Chunk & chunk) const
-    {
-        return getWatermark(chunk, left_watermark_column_position, left_watermark_column_type);
-    }
+        InputPort * input_port;
 
-    inline Int64 getWatermark(const Chunk & chunk, std::optional<size_t> col_pos, const DataTypePtr & col_type) const
-    {
-        if (col_pos)
+        /// Input state
+        /// NOTE: Assume the input chunk is time-ordered
+        /// pair-<retracted_chunk, chunk>
+        SERDE std::list<std::pair<LightChunk, LightChunkWithTimestamp>> input_chunks;
+        /// For join transform, we keep track watermark by itself
+        SERDE Int64 watermark = INVALID_WATERMARK;
+        NO_SERDE Int64 last_data_ts = 0;
+        NO_SERDE CheckpointContextPtr ckpt_ctx = nullptr;
+        NO_SERDE bool muted = false;
+        NO_SERDE bool required_update_processing = false;
+
+        /// Input description
+        std::optional<size_t> watermark_column_position;
+        DataTypePtr watermark_column_type;
+
+        inline void add(Chunk && chunk)
         {
-            auto [_, max_ts] = columnMinMaxTimestamp(chunk.getColumns()[*col_pos], col_type);
-            return max_ts;
+            ckpt_ctx = chunk.getCheckpointContext();
+
+            /// If the input needs to update data, currently the input is always two consecutive chunks with _tp_delta `-1 and +1`
+            /// So we have to process them together before processing another input
+            /// NOTE: Assume the first retracted chunk of updated data always set RetractedDataFlag.
+            if (chunk.isRetractedData())
+            {
+                input_chunks.emplace_back(std::move(chunk), LightChunkWithTimestamp{});
+                required_update_processing = true;
```

Comment:
Look like required_update_processing mud the water a LOT. And this function is too complex (too many branches) and will be slooow.